File: src/main/java/io/github/roguesdev/hoard/DefaultDepositor.java
package io.github.roguesdev.hoard;

import io.github.roguesdev.hoard.serialization.Serializer;
import io.github.roguesdev.hoard.internal.ExceptionUtil;
import io.github.roguesdev.hoard.internal.IOUtils;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.reflect.Type;

class DefaultDepositor<T> implements Depositor<T> {

  private final Serializer serializer;
  private final File saveFile;
  private final Type type;

  DefaultDepositor(Serializer serializer, File directory, String key, Type type) {
    this.serializer = serializer;
    this.saveFile = new File(directory, key);
    this.type = type;
  }

  @Override public synchronized void store(T value) {
    if (value == null) {
      delete();
      return;
    }

    FileOutputStream fileOutputStream = null;
    try {
      fileOutputStream = new FileOutputStream(saveFile);
      serializer.serialize(type, value, fileOutputStream);
    } catch (IOException e) {
      ExceptionUtil.sneakyThrow(e);
    } finally {
      IOUtils.close(fileOutputStream);
    }
  }

  @Override public synchronized T retrieve() {
    if (!saveFile.exists()) {
      return null;
    }

    FileInputStream fileInputStream = null;
    try {
      fileInputStream = new FileInputStream(saveFile);
      return serializer.deserialize(type, fileInputStream);
    } catch (IOException e) {
      ExceptionUtil.sneakyThrow(e);
      return null;
    } finally {
      IOUtils.close(fileInputStream);
    }
  }

  @SuppressWarnings("ResultOfMethodCallIgnored")
  @Override public synchronized void delete() {
    if (!saveFile.exists()) return;
    saveFile.delete();
  }
}

File: src/main/kotlin/io/github/roguesdev/hoard/DefaultDepositor.kt
package io.github.roguesdev.hoard

import io.github.roguesdev.hoard.serialization.Serializer
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.lang.reflect.Type

internal class DefaultDepositor<T>(
  private val serializer: Serializer,
  directory: File,
  key: String,
  private val type: Type
) : Depositor<T> {

  private val saveFile: File = File(directory, key)

  @Synchronized
  override fun store(value: T?) {
    if (value == null) {
      delete()
      return
    }

    return FileOutputStream(saveFile).use {
      serializer.serialize<Any>(type, value, it)
    }
  }

  @Synchronized
  override fun retrieve(): T? {
    if (!saveFile.exists()) {
      return null
    }

    return FileInputStream(saveFile).use {
      serializer.deserialize<T>(type, it)
    }
  }

  @Synchronized
  override fun delete() {
    if (!saveFile.exists()) return
    saveFile.delete()
  }

  @Synchronized
  override fun exists(): Boolean {
    return saveFile.exists()
  }
}
